src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/NoOpRestoreUtilities.cs

```diff
@@ -177,5 +196,48 @@ public static bool VerifyPackagesOnDisk(RestoreRequest request)
             }
             return true;
         }
+
+        /// <summary>
+        /// Calculates the hash value, used for the no-op optimization, for the request
+        /// This methods handles the deduping of tools
+        /// </summary>
+        /// <param name="request"></param>
+        /// <returns></returns>
+        public static string GetHash(RestoreRequest request)
+        {
+            if (request.Project.RestoreMetadata.ProjectStyle == ProjectStyle.DotnetCliTool)
+            {
+                var uniqueName = request.DependencyGraphSpec.Restore.First();
+                var dgSpec = request.DependencyGraphSpec.WithProjectClosure(uniqueName);
+                dgSpec.GetProjectSpec(uniqueName).RestoreMetadata.ProjectPath = null;
+                dgSpec.GetProjectSpec(uniqueName).FilePath = null;
+                return dgSpec.GetHash();
+            }
+
+            return request.DependencyGraphSpec.GetHash();
+        }
+
+        /// <summary>
+        /// This method will resolve the cache/lock file paths for the tool if available in the cache
+        /// </summary>
+        /// <param name="request"></param>
+        public static void ResolveBestMatchingToolPathIfAvailable(RestoreRequest request)
```

Comment:
then this method name is confusing - recommend something like UpdateRestoreRequestToolPathIfAvailable

---
In reply to: [121570248](https://github.com/NuGet/NuGet.Client/pull/1479#discussion_r121570248) [](ancestors = 121570248)